x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/xsearch/action/XSearchAction.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.relevancesearch.xsearch.action;
+
+import org.elasticsearch.action.ActionRequest;
+import org.elasticsearch.action.ActionRequestValidationException;
+import org.elasticsearch.action.ActionType;
+import org.elasticsearch.action.IndicesRequest;
+import org.elasticsearch.action.search.SearchResponse;
+import org.elasticsearch.action.support.IndicesOptions;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Objects;
+
+public class XSearchAction extends ActionType<SearchResponse> {
+
+    public static final XSearchAction INSTANCE = new XSearchAction();
+
+    static final String NAME = "indices:data/read/xsearch";
+
+    private XSearchAction() {
+        super(NAME, SearchResponse::new);
+    }
+
+    public static class Request extends ActionRequest implements IndicesRequest.Replaceable {
+
+        private String[] indices;
+
+        private final String query;
+
+        private final boolean explain;
+        private IndicesOptions indicesOptions = IndicesOptions.fromOptions(false, false, true, false, false, true, false, false);
+
+        public Request(String[] indices, String query, boolean explain) {
+            this.indices = indices;
+            this.query = query;
+            this.explain = explain;
+        }
+
+        public Request(StreamInput in) throws IOException {
+            super(in);
+            this.indices = in.readStringArray();
+            this.query = in.readString();
+            this.explain = in.readBoolean();
+            this.indicesOptions = IndicesOptions.readIndicesOptions(in);
+        }
+
+        public static Request parseRequest(String indexNames, XContentParser bodyParser, boolean explain) throws IOException {
+
+            String[] indices = Strings.splitStringByCommaToArray(indexNames);
+            String query = (String) bodyParser.map().get("query");
+
+            return new Request(indices, query, explain);
+        }
+
+        @Override
+        public ActionRequestValidationException validate() {
+            return null;
```

Comment:
We could do basic validations here, such as making sure user provided non-empty indices, if it's a comma-separated list, then elements are also non-empty. We do something similar here:

https://github.com/elastic/elasticsearch/blob/150f58732b45118e695af31ad6554b6f24d1a660/x-pack/plugin/search-engines/src/main/java/org/elasticsearch/searchengines/action/PutSearchEngineAction.java#L89-L111

Reply:
Good call, adding 👍